sycl/include/sycl/detail/generic_type_traits.hpp

```diff
@@ -328,11 +328,11 @@ template <typename T> auto convertToOpenCLType(T &&x) {
     // TODO: for some mysterious reasons on NonUniformGroups E2E tests fail if
     // we use the "else" version only. I suspect that's an issues with
     // non-uniform groups implementation.
-    if constexpr (std::is_same_v<MatchingVec, no_ref>)
-      return static_cast<typename MatchingVec::vector_t>(x);
-    else
-      return static_cast<typename MatchingVec::vector_t>(
-          x.template as<MatchingVec>());
+    if constexpr (std::is_same_v<MatchingVec, no_ref>) {
+      return bit_cast<typename no_ref::vector_t>(x);
+    } else {
+      return bit_cast<typename MatchingVec::vector_t>(x);
+    }
```

Comment:
If this in fact works, then you can remove the `if constexpr`and just return unconditionally.